In [1]:
import pandas as pd
import numpy as np

In [2]:
print("load & prep data")
print("\treading CSV's")
df_base = pd.read_csv("~/Repositories/datasets/analytics/a_base.csv")
df_qgiv = pd.read_csv("~/Repositories/datasets/analytics/analytics_qgiv.csv")
df_qgiv_base = pd.read_csv("~/Repositories/datasets/analytics/a_qgiv_base_id.csv")

print("\tmerge dataframes")
df_qgiv = df_qgiv.merge(df_qgiv_base, left_on="id", right_on="id_x")
df = df_qgiv.merge(df_base, left_on="base", right_on="id", how="left")

print("\tadding conversion & embedding columns")
df['conversion'] = df['don_form_trans_count'] / (df['visits']+df['mobile_visits'])
df['conversion'].fillna(0., inplace=True)
df['conversion'].replace(np.inf, 1., inplace=True)

load & prep data
	reading CSV's
	merge dataframes
	adding conversion & embedding columns


In [3]:
ftrs_vol = []
ftrs_cnt = []

for c in df.columns:
    if 'vol' in c:
        ftrs_vol.append(c)
    elif 'count' in c:
        ftrs_cnt.append(c)

In [4]:
df[ftrs_vol].corr()

new_rec_volume  new_rec_volume.1  dl_trans_volume  \
new_rec_volume            1.000000          1.000000         0.068684   
new_rec_volume.1          1.000000          1.000000         0.068684   
dl_trans_volume           0.068684          0.068684         1.000000   
dl_new_rec_volume         0.462102          0.462102         0.136096   
vt_trans_vol              0.026612          0.026612         0.001454   
don_form_trans_vol        0.010739          0.010739         0.260960   
kiosk_trans_vol           0.003592          0.003592         0.545129   
p2p_trans_vol                  NaN               NaN              NaN   
mobile_trans_vol          0.012919          0.012919         0.234072   
mobilevt_trans_vol       -0.000018         -0.000018        -0.000094   
sms_trans_vol             0.000899          0.000899         0.042486   
fb_trans_vol              0.001688          0.001688        -0.000118   
one_time_trans_vol        0.010739          0.010739         0.260960   
rec_trans_vol             0.049251          0.049251         0.158523   

                    dl_new_rec_volume  vt_trans_vol  don_form_trans_vol  \
new_rec_volume               0.462102      0.026612            0.010739   
new_rec_volume.1             0.462102      0.026612            0.010739   
dl_trans_volume              0.136096      0.001454            0.260960   
dl_new_rec_volume            1.000000     -0.000065            0.019945   
vt_trans_vol                -0.000065      1.000000            0.017452   
don_form_trans_vol           0.019945      0.017452            1.000000   
kiosk_trans_vol              0.029928     -0.000113            0.015623   
p2p_trans_vol                     NaN           NaN                 NaN   
mobile_trans_vol             0.035481      0.001887            0.150265   
mobilevt_trans_vol          -0.000025      0.408505            0.003243   
sms_trans_vol                0.000883     -0.000093            0.004725   
fb_trans_vol                -0.000031     -0.000033            0.004250   
one_time_trans_vol           0.019945      0.017452            1.000000   
rec_trans_vol                0.275141      0.006457            0.087205   

                    kiosk_trans_vol  p2p_trans_vol  mobile_trans_vol  \
new_rec_volume             0.003592            NaN          0.012919   
new_rec_volume.1           0.003592            NaN          0.012919   
dl_trans_volume            0.545129            NaN          0.234072   
dl_new_rec_volume          0.029928            NaN          0.035481   
vt_trans_vol              -0.000113            NaN          0.001887   
don_form_trans_vol         0.015623            NaN          0.150265   
kiosk_trans_vol            1.000000            NaN          0.054823   
p2p_trans_vol                   NaN            NaN               NaN   
mobile_trans_vol           0.054823            NaN          1.000000   
mobilevt_trans_vol        -0.000057            NaN         -0.000026   
sms_trans_vol              0.061442            NaN          0.010952   
fb_trans_vol              -0.000072            NaN          0.002166   
one_time_trans_vol         0.015623            NaN          0.150265   
rec_trans_vol              0.018678            NaN          0.050604   

                    mobilevt_trans_vol  sms_trans_vol  fb_trans_vol  \
new_rec_volume               -0.000018       0.000899      0.001688   
new_rec_volume.1             -0.000018       0.000899      0.001688   
dl_trans_volume              -0.000094       0.042486     -0.000118   
dl_new_rec_volume            -0.000025       0.000883     -0.000031   
vt_trans_vol                  0.408505      -0.000093     -0.000033   
don_form_trans_vol            0.003243       0.004725      0.004250   
kiosk_trans_vol              -0.000057       0.061442     -0.000072   
p2p_trans_vol                      NaN            NaN           NaN   
mobile_trans_vol             -0.000026       0.010952      0.002166   
mobilev

In [5]:
df[ftrs_cnt].corr()

pledges_count  events_count  events_priv_count  \
pledges_count              1.000000      0.033556           0.074235   
events_count               0.033556      1.000000           0.097077   
events_priv_count          0.074235      0.097077           1.000000   
new_rec_count              0.011199     -0.001345           0.003112   
reg_count                 -0.001436      0.108031           0.042093   
dl_trans_count            -0.005761      0.011951           0.008127   
dl_new_rec_count           0.004205      0.002604           0.007585   
vt_trans_count             0.013128      0.004541           0.001128   
don_form_trans_count      -0.002057      0.024878           0.008186   
kiosk_trans_count         -0.003707      0.012594          -0.002352   
p2p_trans_count                 NaN           NaN                NaN   
mobile_trans_count        -0.005148      0.016394           0.011042   
mobilevt_trans_count       0.001081      0.003686          -0.000420   
sms_trans_count           -0.001777     -0.000697          -0.000204   
fb_trans_count            -0.000804      0.000415          -0.000448   
one_time_trans_count      -0.002058      0.024878           0.008186   
rec_trans_count            0.007905      0.000610           0.004805   

                      new_rec_count  reg_count  dl_trans_count  \
pledges_count              0.011199  -0.001436       -0.005761   
events_count              -0.001345   0.108031        0.011951   
events_priv_count          0.003112   0.042093        0.008127   
new_rec_count              1.000000   0.002056        0.064415   
reg_count                  0.002056   1.000000        0.034816   
dl_trans_count             0.064415   0.034816        1.000000   
dl_new_rec_count           0.274091   0.002717        0.189590   
vt_trans_count             0.218689   0.384623        0.007397   
don_form_trans_count       0.095649   0.215037        0.159303   
kiosk_trans_count          0.005560   0.041712        0.603996   
p2p_trans_count                 NaN        NaN             NaN   
mobile_trans_count         0.122352   0.086594        0.381558   
mobilevt_trans_count      -0.000083   0.028147       -0.000153   
sms_trans_count            0.010069   0.000999        0.062204   
fb_trans_count             0.013106  -0.000164       -0.000152   
one_time_trans_count       0.095649   0.215037        0.159303   
rec_trans_count            0.155392   0.001987        0.053210   

                      dl_new_rec_count  vt_trans_count  don_form_trans_count  \
pledges_count                 0.004205        0.013128             -0.002057   
events_count                  0.002604        0.004541              0.024878   
events_priv_count             0.007585        0.001128              0.008186   
new_rec_count                 0.274091        0.218689              0.095649   
reg_count                     0.002717        0.384623              0.215037   
dl_trans_count                0.189590        0.007397              0.159303   
dl_new_rec_count              1.000000       -0.000203              0.037829   
vt_trans_count               -0.000203        1.000000              0.020859   
don_form_trans_count          0.037829        0.020859              1.000000   
kiosk_trans_count             0.043242       -0.000092              0.018200   
p2p_trans_count                    NaN             NaN                   NaN   
mobile_trans_count            0.135089        0.012018              0.230639   
mobilevt_trans_count         -0.000063        0.031596              0.001302   
sms_trans_count               0.006278        0.000086              0.009784   
fb_trans_count               -0.000063       -0.000055              0.010941   
one_time_trans_count          0.037829        0.020859              1.000000   
rec_trans_count               0.057117        0.015139              0.099072   

                      kiosk_trans_count  p2p_trans_count  mobile_trans_count  \
pledges_count  

In [6]:
df[['new_rec_count', 'dl_new_rec_count']].describe()

new_rec_count  dl_new_rec_count
count   9.000000e+06      9.000000e+06
mean    4.549778e-03      6.528889e-04
std     1.134821e-01      2.772116e-02
min     0.000000e+00      0.000000e+00
25%     0.000000e+00      0.000000e+00
50%     0.000000e+00      0.000000e+00
75%     0.000000e+00      0.000000e+00
max     4.800000e+01      4.000000e+00

In [7]:
df.groupby('enable_donorlogins')[['new_rec_count', 'dl_new_rec_count']].describe()

new_rec_count                                          \
                           count      mean       std  min  25%  50%  75%   
enable_donorlogins                                                         
0                      6382419.0  0.003748  0.107314  0.0  0.0  0.0  0.0   
1                      2617581.0  0.006506  0.127254  0.0  0.0  0.0  0.0   

                         dl_new_rec_count                                     \
                     max            count      mean       std  min  25%  50%   
enable_donorlogins                                                             
0                   36.0        6382419.0  0.000023  0.005236  0.0  0.0  0.0   
1                   48.0        2617581.0  0.002189  0.050715  0.0  0.0  0.0   

                              
                    75%  max  
enable_donorlogins            
0                   0.0  3.0  
1                   0.0  4.0

In [8]:
df.groupby("enable_donorlogins")[['one_time_trans_vol', 'one_time_trans_count']].describe()

one_time_trans_vol                                       \
                                count      mean         std  min  25%  50%   
enable_donorlogins                                                           
0                            288657.0  3.515940   77.054026  0.0  0.0  0.0   
1                            118364.0  5.734587  100.477191  0.0  0.0  0.0   

                                 one_time_trans_count                      \
                    75%      max                count      mean       std   
enable_donorlogins                                                          
0                   0.0  20000.0             288657.0  0.021991  0.403773   
1                   0.0   6643.0             118364.0  0.029705  0.269701   

                                               
                    min  25%  50%  75%    max  
enable_donorlogins                             
0                   0.0  0.0  0.0  0.0  135.0  
1                   0.0  0.0  0.0  0.0   15.0

In [9]:
df.groupby("enable_donorlogins")[['rec_trans_vol', 'rec_trans_count']].describe()

rec_trans_vol                                           \
                           count      mean        std  min  25%  50%  75%   
enable_donorlogins                                                          
0                       288657.0  0.384309  12.227315  0.0  0.0  0.0  0.0   
1                       118364.0  1.062102  26.246094  0.0  0.0  0.0  0.0   

                            rec_trans_count                                \
                        max           count      mean       std  min  25%   
enable_donorlogins                                                          
0                   1750.00        288657.0  0.007466  0.221616  0.0  0.0   
1                   2696.44        118364.0  0.013281  0.222375  0.0  0.0   

                                    
                    50%  75%   max  
enable_donorlogins                  
0                   0.0  0.0  49.0  
1                   0.0  0.0  16.0

In [10]:
df.groupby("enable_donorlogins")[['kiosk_trans_vol', 'kiosk_trans_count']].describe()

kiosk_trans_vol                                          \
                             count     mean        std  min  25%  50%  75%   
enable_donorlogins                                                           
0                         288657.0  0.26474  17.126765  0.0  0.0  0.0  0.0   
1                         118364.0  1.02725  42.241556  0.0  0.0  0.0  0.0   

                            kiosk_trans_count                                \
                        max             count      mean       std  min  25%   
enable_donorlogins                                                            
0                   4519.78          288657.0  0.002643  0.131282  0.0  0.0   
1                   8709.00          118365.0  0.008913  0.287648  0.0  0.0   

                                    
                    50%  75%   max  
enable_donorlogins                  
0                   0.0  0.0  25.0  
1                   0.0  0.0  45.0

In [11]:
df.groupby("enable_donorlogins")[['mobile_trans_vol', 'mobile_trans_count']].describe()

mobile_trans_vol                                           \
                              count      mean        std  min  25%  50%  75%   
enable_donorlogins                                                             
0                          288657.0  0.323209  20.588105  0.0  0.0  0.0  0.0   
1                          118364.0  2.210739  59.867657  0.0  0.0  0.0  0.0   

                           mobile_trans_count                                \
                       max              count      mean       std  min  25%   
enable_donorlogins                                                            
0                   5000.0           288657.0  0.002761  0.109969  0.0  0.0   
1                   6050.0           118365.0  0.011431  0.181694  0.0  0.0   

                                    
                    50%  75%   max  
enable_donorlogins                  
0                   0.0  0.0  19.0  
1                   0.0  0.0  17.0

In [12]:
org_dl_means = df.groupby('org_x')['enable_donorlogins'].mean().reset_index()
org_dl_means.columns = ['org', 'dl']
org_dl_means.head()

org        dl
0    6  1.000000
1    9  0.718607
2   13  0.000000
3   31  0.000000
4   33  1.000000

In [13]:
orgs_dl_changed = org_dl_means[(org_dl_means.dl!=1.)&(org_dl_means.dl!=0.)]
print("{} total orgs, {} orgs that changed dl setting".format(len(df.org_x.unique()), len(orgs_dl_changed)))

2102 total orgs, 331 orgs that changed dl setting


In [14]:
# add column for whether or not the org has changed the setting
df['dl_setting_changed'] = df.org_x.isin(orgs_dl_changed.org.values)

In [15]:
# add column for the last donor login setting
last_dl_settings = df.groupby('org_x')['enable_donorlogins'].last().reset_index()
df['dl_setting_last'] = df['org_x'].apply(lambda x: last_dl_settings[last_dl_settings['org_x']==x]['enable_donorlogins'].values[0])
df[['dl_setting_changed', 'dl_setting_last']].head()

dl_setting_changed  dl_setting_last
0                True                0
1                True                0
2               False                0
3                True                0
4                True                1

In [16]:
cols = ['new_rec_count', 'dl_new_rec_count', 'one_time_trans_vol', 'one_time_trans_count', 'kiosk_trans_vol', 'kiosk_trans_count', 'rec_trans_vol', 'rec_trans_count']
dl_on = df[df.enable_donorlogins==True]
dl_off = df[df.enable_donorlogins==False]

print("| Column | On | Off |")
print("|---|---|---|")
for c in cols:
    print("| {} | {:.6f} | {:.6f} |".format(c, dl_on[c].mean(), dl_off[c].mean()))

| Column | On | Off |
|---|---|---|
| new_rec_count | 0.006506 | 0.003748 |
| dl_new_rec_count | 0.002189 | 0.000023 |
| one_time_trans_vol | 5.734587 | 3.515940 |
| one_time_trans_count | 0.029705 | 0.021991 |
| kiosk_trans_vol | 1.027250 | 0.264740 |
| kiosk_trans_count | 0.008913 | 0.002643 |
| rec_trans_vol | 1.062102 | 0.384309 |
| rec_trans_count | 0.013281 | 0.007466 |


| Column | On | Off |
|---|---|---|
| new_rec_count | 0.006506 | 0.003748 |
| dl_new_rec_count | 0.002189 | 0.000023 |
| one_time_trans_vol | 5.734587 | 3.515940 |
| one_time_trans_count | 0.029705 | 0.021991 |
| kiosk_trans_vol | 1.027250 | 0.264740 |
| kiosk_trans_count | 0.008913 | 0.002643 |
| rec_trans_vol | 1.062102 | 0.384309 |
| rec_trans_count | 0.013281 | 0.007466 |

In [17]:
orgs_dl_on = dl_on['org_x'].unique()

# DL transactions

In [18]:
trans = pd.read_csv("/Users/jeremyvanvalkenburg/Repositories/datasets/transactions/transactions_agg.csv")

In [19]:
dl_on_trans = trans[trans.org.isin(orgs_dl_on)]
dl_off_trans = trans[~trans.org.isin(orgs_dl_on)]

print("{} transactions for on, {} transactions for off".format(len(dl_on_trans), len(dl_off_trans)))
print("${:.2f} mean transaction for on, ${:.2f} mean transaction for off".format(dl_on_trans.amount.mean(), dl_off_trans.amount.mean()))
print("${:.2f} average funds raised for dl on".format(dl_on_trans.amount.sum() / float(len(orgs_dl_on))))
print("${:.2f} average funds raised for dl off".format(dl_off_trans.amount.sum() / float(len(dl_off_trans.org.unique()))))

2210966 transactions for on, 2322804 transactions for off
$4368.06 mean transaction for on, $1541.12 mean transaction for off
$13469513.99 average funds raised for dl on
$1279386.51 average funds raised for dl off


In [20]:
trans.amount.mean()

2919.7257440521357